### PR TITLE
The ring `ℤ/n` is cyclic for any natural number `n`

### PR DESCRIPTION
**Pull request by GregorPercic**
In this addition, we have proven the following.
- If a function `f` is homotopic to a function `g` and `g` is surjective, `f` is surjective (`is-surjective-htpy`).
- `mod-ℤ` is surjective.
- `mod-ℤ` and multiplication of 1 by an integer in the ring ℤ/n are homotopic.
- ℤ/n is a cyclic ring, for any natural number n.

### CODE REVIEW

**Comment by EgbertRijke** [comment]
Awesome! Please update your branch:
- do a `git pull upstream master` or pull `upstream master` from within VScode to update your branch locally
- Commit changes if necessary
- push

Then your changes should automatically show up here in the pull request.

We can also do it from here, if you have any trouble with it, but this will cause the PR to be out of sync with your local branch, so I would prefer the first method. Let me know if you have any trouble.

File: .vscode/settings.json [hunk 1]
```diff
@@ -56,6 +56,8 @@
 
   // File type specific
   "files.associations": {
+    "*.agda": "agda",
+    "*.lagda.md": "lagda-md",
```
Comment (EgbertRijke) [inline .vscode/settings.json:60]: This shouldn't be necessary anymore. @fredrik-bakke ?

File: src/elementary-number-theory/modular-arithmetic.lagda.md [hunk 2]
```diff
@@ -788,3 +791,23 @@ is-decidable-div-ℤ d x =
       ( mod-ℤ (abs-ℤ d) x)
       ( zero-ℤ-Mod (abs-ℤ d)))
 ```
+
+### `mod-ℤ` is surjective
+
+```agda
+is-surjective-succ-Fin-comp-mod-succ-ℕ : (n : ℕ) →
+  is-surjective (succ-Fin (succ-ℕ n) ∘ mod-succ-ℕ n)
```
Comment (EgbertRijke) [inline src/elementary-number-theory/modular-arithmetic.lagda.md:799]:
```suggestion
is-surjective-succ-Fin-comp-mod-succ-ℕ :
  (n : ℕ) → is-surjective (succ-Fin (succ-ℕ n) ∘ mod-succ-ℕ n)
```

We make a line break immediately after the colon if the entire specification doesn't fit on a line with the name.

File: src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md [hunk 3]
```diff
@@ -53,3 +65,120 @@ pr2 (pr2 (pr2 (pr2 (ℤ-Mod-Ring n)))) = right-distributive-mul-add-ℤ-Mod n
...
+    ( ℤ-Mod-Group n)
+    ( one-ℤ-Mod n)
+    ( is-surjective-hom-element-one-ℤ-Mod-Ring n)
+
+is-cyclic-ℤ-Mod-Ring :
+  ( n : ℕ) → is-cyclic-Ring (ℤ-Mod-Ring n)
+is-cyclic-ℤ-Mod-Ring n =
+  intro-∃
+    ( one-ℤ-Mod n)
+    ( is-generating-element-one-ℤ-Mod-Ring n)
```
Comment (EgbertRijke) [inline src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md:183]: Awesome!

File: src/ring-theory/integer-multiples-of-elements-rings.lagda.md [hunk 4]
```diff
@@ -206,7 +206,7 @@ module _
     integer-multiple-neg-Ab (ab-Ring R)
 ```
 
-### `(k + 1)x = kx + x` and `(k+1)x = x + kx`
+### `(k + 1)x = kx + x` and `(k + 1)x = x + kx`
```
Comment (EgbertRijke) [inline src/ring-theory/integer-multiples-of-elements-rings.lagda.md:209]: Well spotted!

**Review by EgbertRijke — APPROVED** [review]
Very nice pull request! I'm very happy with this result, well done!

The file `rings-of-modular-arithmetic` might benefit from some reorganization, but I'm happy to do that myself after the merge. I don't have any additional changes to request.

File: .vscode/settings.json (hunk 1 above)
Comment (fredrik-bakke) [reply .vscode/settings.json:60]: no, that must be removed

Comment (fredrik-bakke) [reply .vscode/settings.json:60]: Thanks for spotting it!

Comment (EgbertRijke) [inline .vscode/settings.json:60]:
```suggestion
```

File: src/ring-theory/rings.lagda.md [hunk 5]
```diff
@@ -334,6 +334,16 @@ module _
           ( right-inverse-law-add-Ring y))
       ＝ add-Ring R x z
         by left-unit-law-add-Ring R (add-Ring R x z)
+
+  eq-is-unit-left-div-Ring : {x y : type-Ring R} →
```
Comment (fredrik-bakke) [inline src/ring-theory/rings.lagda.md:338]:
```suggestion
  eq-is-unit-left-div-Ring :
    {x y : type-Ring R} →
```

File: src/ring-theory/rings.lagda.md [hunk 6]
```diff
@@ -334,6 +334,16 @@ module _
...
+    (is-zero-Ring R (add-Ring R (neg-Ring x) y)) → x ＝ y
+  eq-is-unit-left-div-Ring {x} {y} H =
+    eq-is-unit-left-div-Group (group-Ring R) H
+
+  is-unit-left-div-eq-Ring : {x y : type-Ring R} →
```
Comment (fredrik-bakke) [inline src/ring-theory/rings.lagda.md:343]:
```suggestion
  is-unit-left-div-eq-Ring :
    {x y : type-Ring R} →
```

File: .vscode/settings.json (hunk 1 above)
Comment (GregorPercic) [reply .vscode/settings.json:60]: Yep, that was my temporary "patch-fix." I am surprised it is still there, though, as I have updated this branch after Fredrik added the official solution in the main repository.

File: src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md [hunk 7]
```diff
@@ -53,3 +65,120 @@ pr2 (pr2 (pr2 (pr2 (ℤ-Mod-Ring n)))) = right-distributive-mul-add-ℤ-Mod n
 pr1 (ℤ-Mod-Commutative-Ring n) = ℤ-Mod-Ring n
 pr2 (ℤ-Mod-Commutative-Ring n) = commutative-mul-ℤ-Mod n
 ```
+
+## Properties
+
+### Rings of modular arithmetic are cyclic
+
+```agda
+integer-multiple-one-ℤ-Ring : (k : ℤ) →
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md:74]:
```suggestion
integer-multiple-one-ℤ-Ring :
  (k : ℤ) →
```

File: src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md [hunk 8]
```diff
@@ -53,3 +65,120 @@ pr2 (pr2 (pr2 (pr2 (ℤ-Mod-Ring n)))) = right-distributive-mul-add-ℤ-Mod n
...
+integer-multiple-one-ℤ-Ring (inr (inr zero-ℕ)) = refl
+integer-multiple-one-ℤ-Ring (inr (inr (succ-ℕ n))) =
+  ap succ-ℤ (integer-multiple-one-ℤ-Ring (inr (inr n)))
+
+is-neg-one-neg-one-ℤ-Mod :
+  ( n : ℕ) → (neg-one-Ring (ℤ-Mod-Ring n)) ＝ (neg-one-ℤ-Mod n)
+is-neg-one-neg-one-ℤ-Mod zero-ℕ = refl
+is-neg-one-neg-one-ℤ-Mod (succ-ℕ n) = is-neg-one-neg-one-Fin n
+
+integer-multiplication-by-one-preserves-succ-ℤ : (n : ℕ) (x : ℤ) →
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md:90]:
```suggestion
integer-multiplication-by-one-preserves-succ-ℤ :
  (n : ℕ) (x : ℤ) →
```

File: src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md [hunk 9]
```diff
@@ -53,3 +65,120 @@ pr2 (pr2 (pr2 (pr2 (ℤ-Mod-Ring n)))) = right-distributive-mul-add-ℤ-Mod n
...
+  integer-multiple-Ring (ℤ-Mod-Ring n) (succ-ℤ x) (one-ℤ-Mod n) ＝
+  succ-ℤ-Mod n (integer-multiple-Ring (ℤ-Mod-Ring n) x (one-ℤ-Mod n))
+integer-multiplication-by-one-preserves-succ-ℤ n x =
+  ( integer-multiple-succ-Ring (ℤ-Mod-Ring n) x (one-ℤ-Mod n)) ∙
+  ( inv
+    ( is-left-add-one-succ-ℤ-Mod'
+      ( n)
+      ( integer-multiple-Ring (ℤ-Mod-Ring n) x (one-ℤ-Mod n))))
+
+integer-multiplication-by-one-preserves-pred-ℤ : (n : ℕ) (x : ℤ) →
```
Comment (fredrik-bakke) [inline src/elementary-number-theory/rings-of-modular-arithmetic.lagda.md:100]:
```suggestion
integer-multiplication-by-one-preserves-pred-ℤ :
  (n : ℕ) (x : ℤ) →
```

**Review by fredrik-bakke — APPROVED** [review]
I must say your PR is very well written!

**Comment by fredrik-bakke** [comment]
The "Resolve conversation" button only flags the comment letting others know you consider the issue fixed. It does not change the code. However, for most of our comments here, we've provided suggested changes to fix the issues. These you can commit directly from the GitHub user interface by clicking this button:
Screenshot 2023-09-29 at 19 27 02

**Comment by EgbertRijke** [comment]
I went ahead and committed them all

**Comment by EgbertRijke** [comment]
Well done! Thank you again for this pull request! We will merge it when the checks go green.

👌

**Pull request merged**